src/treelearner/kernels/histogram_16_64_256.cu

```diff
@@ -0,0 +1,1008 @@
+        #endif
+
+        // imbGBT v5.1
+        grad_next = ordered_gradients[ind_next];
+        #if CONST_HESSIAN == 0
+        hess_next = ordered_hessians[ind_next];
+        #endif
+
+        // STAGE 2: accumulate gradient and hessian
+        if (bin != feature) {
+            addr_bin = gh_hist + bin * 2 + is_hessian_first;
+            #if CONST_HESSIAN == 0
+            acc_type acc_bin = is_hessian_first ? hess_bin : grad_bin;
+            atomic_local_add_f(addr_bin, acc_bin);
+
+            addr_bin = addr_bin + 1 - 2 * is_hessian_first;
+            acc_bin = is_hessian_first ? grad_bin : hess_bin;
+            atomic_local_add_f(addr_bin, acc_bin);
+
+            #elif CONST_HESSIAN == 1
+            atomic_local_add_f(addr_bin, grad_bin);
+            #endif
+
+            bin = feature;
+            grad_bin = grad;
+            hess_bin = hess;
+        } else {
+            grad_bin += grad;
+            hess_bin += hess;
+        }
+
+        // prefetch the next iteration variables
+        feature_next = feature_data[ind_next >> feature_mask];
+
+        // STAGE 3: accumulate counter
+        atomicAdd(cnt_hist + feature, 1);
+
+        // STAGE 4: update next stat
+        grad = grad_next;
+        hess = hess_next;
+        // LGBM_CUDA: v4.2
+        if (!feature_mask) {
+            feature = feature_next;
+        } else {
+            feature = (feature_next >> ((ind_next & 1) << 2)) & 0xf;
+        }
+    }
+
+
+    addr_bin = gh_hist + bin * 2 + is_hessian_first;
+    #if CONST_HESSIAN == 0
+    acc_type acc_bin = is_hessian_first ? hess_bin : grad_bin;
+    atomic_local_add_f(addr_bin, acc_bin);
+
+    addr_bin = addr_bin + 1 - 2 * is_hessian_first;
+    acc_bin = is_hessian_first ? grad_bin : hess_bin;
+    atomic_local_add_f(addr_bin, acc_bin);
+
+    #elif CONST_HESSIAN == 1
+    atomic_local_add_f(addr_bin, grad_bin);
+    #endif
+    __syncthreads();
+
+    #if CONST_HESSIAN == 1
+    // make a final reduction
+    gh_hist[ltid * 2] += gh_hist[ltid * 2 + 1];
+    gh_hist[ltid * 2 + 1] = const_hessian * cnt_hist[ltid];  // LGBM_CUDA: counter move to this position
+    __syncthreads();
+    #endif
+
+#if POWER_FEATURE_WORKGROUPS != 0
+    acc_type *__restrict__ output = reinterpret_cast<acc_type *>(output_buf) + group_id * 3 * NUM_BINS;
+    // write gradients and hessians
+    acc_type *__restrict__ ptr_f = output;
+    for (ushort i = ltid; i < 2 * NUM_BINS; i += lsize) {
+        // even threads read gradients, odd threads read hessians
+        acc_type value = gh_hist[i];
+        ptr_f[(i & 1) * NUM_BINS + (i >> 1)] = value;
+    }
+    // write counts
+    acc_int_type *__restrict__ ptr_i = reinterpret_cast<acc_int_type *>(output + 2 * NUM_BINS);
+    for (ushort i = ltid; i < NUM_BINS; i += lsize) {
+        uint value = cnt_hist[i];
+        ptr_i[i] = value;
+    }
+    __syncthreads();
+    __threadfence();
+    // To avoid the cost of an extra reducting kernel, we have to deal with some
```

Comment:
These comments (line 297-309; line 627-639) can be removed.
The `asm volatile` is the same as `__threadfence()` so can be removed.